app/core/store/modules/unitMap.js

```diff
@@ -0,0 +1,84 @@
+import api from 'core/api'
+
+export default {
+  namespaced: true,
+  state: {
+    currentLevelsList: {}
+  },
+
+  mutations: {
+    setCurrentLevelsList: (state, levels) => {
+      state.currentLevelsList = levels
+    }
+  },
+
+  getters: {
+    getCurrentLevelsList: (state) => state.currentLevelsList
+  },
+
+  actions: {
+    /**
+      We have a campaign.levels list and a classroom.courses.levels list, and they are not always in sync.
+      Hence to get the levels data for a course instance for the unit map, we get the data as follows:
+      1. levels list from the classroom snapshot
+      2. position, nextLevels, first from the classroom snapshot, if does not exist then from campaign snapshot
+      4. any other data from the campaign snapshot, but if doesnt exist in campaign any more then use the data in classroom snapshot
+      */
+    buildLevelsData: async ({ commit, rootGetters, dispatch }, campaignHandle, courseInstanceId) => { // campaignHandle necessary
+      let levels = {}
+      await dispatch('campaigns/fetch', campaignHandle, { root: true })
+      const campaignData = rootGetters['campaigns/getCampaignData'](campaignHandle)
+      if (!courseInstanceId) {
+        levels = campaignData.levels
+      } else {
+        try {
+          const courseInstance = await api.courseInstances.get({ courseInstanceID: courseInstanceId })
+          const courseId = courseInstance.courseID
+          const classroomId = courseInstance.classroomID
+
+          // campaign snapshot of the levels
+          const existingCampaignLevels = _.cloneDeep(campaignData.levels)
+
+          // classroom snapshot of the levels for the course
+          const classroom = await api.classrooms.get({ classroomID: classroomId })
+          const classroomCourseLevels = _.find(classroom.courses, { _id: courseId }).levels
+
+          // get levels data for the levels in the classroom snapshot
+          const classroomCourseLevelsData = await api.classrooms.getCourseLevels({ classroomID: classroomId, courseID: courseId })
```

Comment:
How expensive is it to make 3 back and forth requests. When is it more efficient to make a new endpoint that wraps the back and forth exchange into one request?
Just a random thought, and requires no action.